bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java

```diff
@@ -0,0 +1,478 @@
+        } else if (tivoSocket.isClosed()) {
+            logger.debug(" connIsConnected '{}' - FALSE: tivoSocket.isClosed=true", tivoConfigData.getCfgIdentifier());
+            return false;
+        } else if (streamWriter == null) {
+            logger.debug(" connIsConnected '{}' - FALSE: tivoIOSendCommand=null", tivoConfigData.getCfgIdentifier());
+            return false;
+        } else if (streamWriter.checkError()) {
+            logger.debug(" connIsConnected '{}' - FALSE: tivoIOSendCommand.checkError()=true",
+                    tivoConfigData.getCfgIdentifier());
+            return false;
+        } else if (streamReader == null) {
+            logger.debug(" connIsConnected '{}' - FALSE: streamReader=null", tivoConfigData.getCfgIdentifier());
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * {@link connTivoConnect} manages the creation / retry process of the socket connection.
+     *
+     * @return true = connected, false = not connected
+     * @throws InterruptedException
+     */
+    public boolean connTivoConnect() throws InterruptedException {
+        for (int iL = 1; iL <= tivoConfigData.getNumRetry(); iL++) {
+            logger.debug(" connTivoConnect '{}' - starting connection process '{}' of '{}'.",
+                    tivoConfigData.getCfgIdentifier(), iL, tivoConfigData.getNumRetry());
+
+            // Sort out the socket connection
+            if (connSocketConnect()) {
+                logger.debug(" connTivoConnect '{}' - Socket created / connection made.",
+                        tivoConfigData.getCfgIdentifier());
+                StreamReader streamReader = this.streamReader;
+                if (streamReader != null && streamReader.isAlive()) {
+                    return true;
+                }
+            } else {
+                logger.debug(" connTivoConnect '{}' - Socket creation failed.", tivoConfigData.getCfgIdentifier());
+                TiVoHandler tivoHandler = this.tivoHandler;
+                if (tivoHandler != null) {
+                    tivoHandler.setStatusOffline();
+                }
+            }
+            // Sleep and retry
+            doNappTime();
+        }
+        return false;
+    }
+
+    /**
+     * {@link connTivoDisconnect} conditionally closes the Socket connection. When 'keep connection open' or 'channel
+     * scanning' is true, the disconnection process is ignored. Disconnect can be forced by setting forceDisconnect to
+     * true.
+     *
+     * @param forceDisconnect true = forces a disconnection , false = disconnects in specific situations
+     * @throws InterruptedException
+     */
+    public void connTivoDisconnect(boolean forceDisconnect) throws InterruptedException {
+        if (forceDisconnect) {
+            connSocketDisconnect();
+        } else {
+            if (!tivoConfigData.isKeepConnActive()) {
+                doNappTime();
+                connSocketDisconnect();
+            }
+        }
```

Comment:
You are performing two different behaviors based on the input boolean parameter. It would be better if you split this into two different methods instead.

Reply:
Agreed. I simplified the logic.